src/fides/api/ops/db/base_class.py

```diff
@@ -296,15 +294,8 @@ def save(self, db: Session) -> FidesBase:
         """Save the current object over an existing row in the database."""
         if hasattr(self, "key"):
             key = getattr(self, "key")
-
-            is_valid = False
-            if key is not None:
-                is_valid = len(ALLOWED_CHARS.sub("", key)) == 0
-
-            if not is_valid:
-                raise KeyValidationError(
-                    f"Key '{key}' on {self.__class__.__name__} is invalid."
-                )
+            assert key, f"Key on class '{self.__class__.__name__}' cannot be empty!"
```

Comment:
can we remove the assert and throw a proper error?

Reply:
i too thought this looked funny but figured it may be OK in this sort of context? but if you're also mentioning it, then seems like it's worth addressing!

Reply:
When to use `assert` vs a general `exception` is very much a subjective matter, but for me the guidelines are generally:

- Exceptions: Use an `exception` when the error would be caused by input from a 3rd-party, such as user input or API call
- Assertion: Use when the "call is coming from inside the house", i.e. your code itself created the input that would allow the program to get into a bad state and you want to catch that early and fail fast.

To confirm, there is no validation being done before this step? Such as when Pydantic/FastAPI validate the payload of the incoming request?

I'm by no means going to staunchly defending using an `assert` here, but based on my understanding of the code and generally how I employ asserts it made sense to me here. I also could be making an incorrect assumption about the code up to this point though, and that this really is the only point of validation

Reply:
Pydantic is validating the request but this step was previously getting hit for various characters discussed below https://github.com/ethyca/fides/pull/3312#discussion_r1196965142 (which also seems like something we should have been catching and throwing a proper HTTP Exception in the endpoint.)  However, I don't expect this particular assert to be hit.